### ISSUE TITLE
FSDP memory utilization issue 

### ISSUE THREAD

**Issue by prigoyal** [facebookresearch/fairscale#627]
Reporting the issue here:

We observe weird memory usage when using FSDP. 

**Setup:** 1.5B params model, no activation checkpointing used

**Observations are:**
- **DDP, 1gpu, 1img/gpu , Pytorch AMP , memory looks like:**
    - 5480 (before fwd) -> 9547 (after fwd) -> 10960 (after bwd) -> 16384 (after optimizer step and params update)
    - verified with PyTorch 1.6 and PyTorch 1.9.0 both have same

- **FSDP (newest integration), 2gpu, 1img/gpu, compute_dtype=fp32 and equivalent to O1 settings, memory looks like**
    - 2739 (before fwd) -> 7416 (after fwd) -> **11190 (after bwd)** -> 12599 (after optimizer step and params update)
    - verified with Pytorch 1.6 and Pytorch 1.9.0 both
    - **What's wrong**: On backward we would expect the memory to be 2739 (params sharded) + 2739 (gradients sharded) but instead we see 11190

- **FSDP (integration from March beginning that was studied for perf), 2gpu, 1img/gpu, compute_dtype=fp32 and equivalent to O1 settings, memory looks like**
    - 274 (before fwd) -> 7412 (after fwd) -> **5488 (after bwd)** -> 5774 (after optimizer step and params update)
    - Pytorch 1.9.0 from March

**_Something has changed in the newest integration and after backward pass with FSDP, we don't see the expected memory_**.

**Things we have ruled out:**

- [x] Pytorch memory fragmentation: 
- [x] Pytorch version issue
- [x] Issue with FSDP -> **CONFIRMED**
- [x] Issue in our integration

# Some more details on FSDP vs DDP memory consumption

## In the context of VISSL

We have been noticing that for RegNet16, the memory consumption with FSDP is significantly higher than DDP. The key differences happen in iteration 0. All the data below is obtains on a RegNet16, trained on 8 GPUs for a few epochs with a batch size of 2 (small batch size to limit the contribution of activation memory).

### At iteration 0

Here is a graph of the memory consumption during forward/backward **in DPP**:

<img width="523" alt="Screenshot 2021-04-23 at 11 44 01" src="https://user-images.githubusercontent.com/7412790/115896278-400adc00-a429-11eb-835f-80308a58fb47.png">

Here is a graph of the memory consumption during forward/backward **in FSDP**:

<img width="529" alt="Screenshot 2021-04-23 at 11 44 12" src="https://user-images.githubusercontent.com/7412790/115896285-413c0900-a429-11eb-9fcb-aa6c61a34b2e.png">

The key difference is the "tilt": after the end of the forward/backward, FSDP retrains a lot more memory than DDP, although it is expected that it retains less memory: the optimizer state as well as the gradients are sharded and should be smaller.

### At iteration 1

This tilt at iteration 0 is problematic because it means that some memory is never freed, and indeed, starting from iteration 1, FSDP memory usage is significantly higher because of this "base memory usage".

Here is the curve for DDP during forward/backward:

<img width="515" alt="Screenshot 2021-04-23 at 11 48 17" src="https://user-images.githubusercontent.com/7412790/115896854-d808c580-a429-11eb-833c-34c79cc79469.png">

Here is the curve for FSDP during forward/backward:

<img width="525" alt="Screenshot 2021-04-23 at 11 48 27" src="https://user-images.githubusercontent.com/7412790/115896864-d9d28900-a429-11eb-87c8-59a2433b7691.png">

More memory is used in FSDP from this iteration on, because some memory has been retained from the first iteration, and this memory makes FSDP consumes 1 GB more memory although the model is sharded across 8 GPUs.

## On a custom model (unit test)

Similar memory consumption curves can be reproduced on simpler model, not tied to VISSL (the script is given below):
- a conv net with just a few layers
- no optimizer states (simple SGD)
- only 2 GPUs used

Here we compare the memory consumption of FSDP vs DDP on the 5 first iterations. The memory does not goes down as steeply with FSDP:

<img width="604" alt="Screenshot 2021-04-23 at 11 42 09" src="https://user-images.githubusercontent.com/7412790/115895966-f7ebb980-a428-11eb-8ff4-2cd76a916947.png">

The code to reproduce the issue (in comments are the lines used to fetch the memory curves, this tooling is available in this branch https://github.com/fairinternal/ssl_scaling/pull/100 if you are interested)

```
import argparse
import pickle
from typing import Union, List

import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.nn as nn
import torch.optim as optim
from fairscale.nn import checkpoint_wrapper
from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP, auto_wrap_bn
from torch.nn.parallel import DistributedDataParallel
# from vissl.utils.layer_memory_tracking import LayerwiseMemoryTracker

def get_global_group():
    """
    Singleton pytorch distributed group
    Inspired by https://github.com/pytorch/fairseq
    """
    if dist.is_initialized():
        if not hasattr(get_global_group, "_global_group"):
            get_global_group._global_group = dist.new_group()
        return get_global_group._global_group
    else:
        return None


def to_fsdp(module):
    return FSDP(module, process_group=get_global_group())


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.stem = nn.Sequential(
            nn.Conv2d(3, 64, kernel_size=3),
            nn.BatchNorm2d(64),
            nn.ReLU(inplace=True)
        )
        self.blocks = nn.Sequential(
            nn.Conv2d(64, 128, kernel_size=5, padding=2),
            nn.BatchNorm2d(128),
            nn.ReLU(inplace=True),
            nn.Conv2d(128, 128, kernel_size=5, padding=2),
            nn.BatchNorm2d(128),
            nn.ReLU(inplace=True),
            nn.Conv2d(128, 128, kernel_size=5, padding=2),
            nn.BatchNorm2d(128),
            nn.ReLU(inplace=True),
            nn.AdaptiveAvgPool2d(output_size=(1, 1)),
            nn.Flatten(),
        )
        self.head = nn.Linear(128, 10)

    def forward(self, x: Union[torch.Tensor, List[torch.Tensor]]):
        return self.head(self.blocks(self.stem(x)))


def create_model(with_fsdp: bool, with_checkpoint: bool = False):
    model = Model()
    if with_fsdp:
        model.stem = auto_wrap_bn(model.stem, single_rank_pg=False)
        model.blocks = auto_wrap_bn(model.blocks, single_rank_pg=False)
        if with_checkpoint:
            model.blocks = checkpoint_wrapper(model.blocks)
        model.stem = to_fsdp(model.stem)
        model.blocks = to_fsdp(model.blocks)
        model.head = to_fsdp(model.head)
    else:
        if with_checkpoint:
            model.blocks = checkpoint_wrapper(model.blocks)
    return model


def _distributed_worker(gpu_id: int, with_fsdp: bool):
    torch.cuda.set_device(gpu_id)
    dist.init_process_group(
        backend="nccl", init_method="tcp://127.0.0.1:9099", world_size=2, rank=gpu_id
    )

    torch.manual_seed(0)
    torch.backends.cudnn.deterministic = True
    batch = torch.randn(size=(2, 3, 224, 224)).cuda()

    model = create_model(with_fsdp)
    model = model.cuda()
    if with_fsdp:
        model = to_fsdp(model)
        print(model.local_state_dict().keys())
    else:
        model = DistributedDataParallel(model, device_ids=[gpu_id])

    criterion = nn.MSELoss()
    optimizer = optim.SGD(model.parameters(), lr=1e-4)

    # memory_tracker = LayerwiseMemoryTracker()
    # memory_tracker.monitor(model)

    for iteration in range(5):
        out = model(batch)
        out = sum(o.sum() for o in out[0])
        fake_loss = criterion(out, torch.tensor(0.0).cuda())
        if gpu_id == 0:
            print("Loss", iteration, ":", fake_loss.item())
        optimizer.zero_grad()
        fake_loss.backward()
        optimizer.step()

    if gpu_id == 0:
        # image = memory_tracker.show_plots(capture=True)
        # image.save("path_to_where_to_save_data")
        pass


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-f", "--fsdp", action="store_const", const=True, default=False)
    args = parser.parse_args()
    mp.spawn(_distributed_worker, (args.fsdp,), nprocs=2)
```

**Comment by min-xu-ai** [comment]
The draft PR #633 has part of the fix. @QuentinDuval , VISSL side needs to:

1. get the new version of FSDP once this PR is merged. This enables less memory used after backward is done.
2. go over all FSDP wrapper in vissl to set their bucket size using some kind of gut feeling (:-)). By default every one instance is ~25MB and BN will have 0MB. But you may want to vary that based on the layers being wrapped.
3. Finally, it is important to use `for p in model.parameters(): p.grad=None` with FSDP. Using optimizer.zero_grad() will not free memory as much as setting None for FSDP. The exact reason is not yet clear to me. It is likely related to hooks we placed on the autograd graph.

**Comment by prigoyal** [comment]
thank you for the update. We will test the fix once this PR is merged and will update the issue with the findings. 

I believe that we would set bucket size = 0 by default for all to start and thank you for pointing out the `p.grad=None`. We will ensure the grads are set to None explicitly.

**Comment by min-xu-ai** [comment]
This particular issue is fixed. FSDP now saves more memory outside of the backward pass. User still needs to ensure bucket buffer sizes are set correctly and it is better to use a global process group to avoid duplicated buffer and `model.zero_grad(set_to_none=True)` is highly recommend after `optimizer.step()`.

We will continue debug the 12GB card OOM issue with 1.5B model, but please use a separate issue for better tracking purposes.

**Comment by prigoyal** [comment]
thank you @min-xu-ai , sounds great.

**Comment by s-reaungamornrat** [comment]
Hi All,

I converted my model training from DataParallel to FSDP based on PyTorch FSDP document and its example in (https://github.com/pytorch/examples/blob/main/distributed/FSDP), but I faced the issue that FSDP training requires more memory than DataParallel one. Could anyone provide the link to the documentation clarifying the buffer's cost and speed/memory tradeoff mentioned in https://github.com/facebookresearch/fairscale/pull/633? What is buffer bucket and how can I set its size in FSDP? Thank you very much for your help!
